### PR TITLE
[query] test compiled, unoptimized

### PR DESCRIPTION
**Pull request by cseed**
fix some ptype inference bugs

### CODE REVIEW

File: hail/src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala [hunk 1]
```diff
@@ -139,12 +139,16 @@ class StagedRegionValueBuilder private(val mb: MethodBuilder, val typ: PType, va
...
     ftype match {
-      case t: PArray => t.setElementMissing(startOffset, idx)
-      case t: PCanonicalBaseStruct =>
-        if (t.fieldRequired(staticIdx))
-          Code._fatal[Unit](s"Required field cannot be missing: $t, $staticIdx")
+      case t: PArray =>
+        if (!t.elementType.required)
+          t.setElementMissing(startOffset, idx)
         else
+          Code._fatal[Unit](s"Required field cannot be missing")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala]: the messages are swapped here -- field with array, element with struct

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala [hunk 2]
```diff
@@ -6,9 +6,13 @@ import is.hail.utils._
 
 object InferPType {
 
-  def clearPTypes(x: IR): Unit = {
-    x._pType = null
-    x.children.foreach { c => clearPTypes(c.asInstanceOf[IR]) }
+  def clearPTypes(x: BaseIR): Unit = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: why is this necessary? When we have a ptype-d IR, it must contain only IR, not BaseIR.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 3]
```diff
@@ -104,19 +104,26 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
-    assert(!elementType.required)
-    Region.setBit(aoff + lengthHeaderBytes, i)
+    if (!elementType.required)
+      Region.setBit(aoff + lengthHeaderBytes, i)
   }
 
   def setElementMissing(aoff: Code[Long], i: Code[Int]): Code[Unit] =
-    Region.setBit(aoff + lengthHeaderBytes, i.toL)
+    if (elementType.required)
+      Code._empty
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: should the fatal be here, not on srvb?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: alternatively, could `assert(!elementType.required)`?

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 4]
```diff
@@ -104,19 +104,26 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
     !isElementDefined(aoff, i)
 
   def setElementMissing(aoff: Long, i: Int) {
-    assert(!elementType.required)
-    Region.setBit(aoff + lengthHeaderBytes, i)
+    if (!elementType.required)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: same as below.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 5]
```diff
@@ -433,7 +442,7 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
   }
 
   private def constructOrCopy(mb: MethodBuilder, region: Code[Region], srcArray: PArray, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
-    if (srcArray == this) {
+    if (srcArray.elementType == elementType) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: I see why you made this change (avoid unnecessary copies) but it makes me nervous -- when we add another implementation of PArray, this will break.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: We should add a method `sameExceptTopLevelRequiredness` or something, which is what this is trying to ask, yes?

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala (hunk 2 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: So where I run (inside TestUtils.assertEvalsTo), there can be relational nodes.  I have to do this because we run multiple compile strategies now (opt and non-opt).  Those IR can be shared arbitrarily with IR that get ptypes assigned by InferPTypes.  InferPTypes asserts no ptypes are already assigned, so they all need to be cleared.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala (hunk 5 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: Yes, broken.  I added an `isInstanceOf[PCanonicalArray]`.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala (hunk 3 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: Good, yeah, I moved the assert.  I also did the same thing for PBaseStruct.

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: got it, sure.
